crates/tx5/src/ep3.rs

```diff
@@ -0,0 +1,600 @@
+            ._sig_map
+            .lock()
+            .unwrap()
+            .values()
+            .map(|v| v.1.clone())
+            .collect::<Vec<_>>();
+
+        for fut in fut_list {
+            task_list.push(async move {
+                if let Ok(sig) = fut.await {
+                    Some(sig.get_stats().await)
+                } else {
+                    None
+                }
+            });
+        }
+
+        let res: Vec<(Id, serde_json::Value)> =
+            futures::future::join_all(task_list)
+                .await
+                .into_iter()
+                .flatten()
+                .flatten()
+                .collect();
+
+        let mut map = serde_json::Map::default();
+
+        #[cfg(feature = "backend-go-pion")]
+        const BACKEND: &str = "go-pion";
+        #[cfg(feature = "backend-webrtc-rs")]
+        const BACKEND: &str = "webrtc-rs";
+
+        map.insert("backend".into(), BACKEND.into());
+        map.insert("thisId".into(), self.ep.this_id.to_string().into());
+        map.insert("banned".into(), ban_map.into());
+
+        for (id, v) in res {
+            map.insert(id.to_string(), v);
+        }
+
+        serde_json::Value::Object(map)
+    }
+}
+
+async fn assert_sig(ep: &Arc<EpShared>, sig_url: &SigUrl) -> CRes<Arc<Sig>> {
+    let sig_map = match ep.weak_sig_map.upgrade() {
+        Some(sig_map) => sig_map,
+        None => {
+            return Err(Error::str(
+                "Signal connection failed due to closed endpoint",
+            )
+            .into())
+        }
+    };
+
+    let (sig_uniq, fut) = sig_map
+        .lock()
+        .unwrap()
+        .entry(sig_url.clone())
+        .or_insert_with(|| {
+            let sig_uniq = next_uniq();
+            let sig_url = sig_url.clone();
+            let ep = ep.clone();
+            let _sig_drop = SigDrop {
+                ep_uniq: ep.ep_uniq,
+                sig_uniq,
+                sig_url: sig_url.clone(),
+                weak_sig_map: ep.weak_sig_map.clone(),
+            };
+            (
+                sig_uniq,
+                AbortableTimedSharedFuture::new(
+                    ep.config.timeout,
+                    Error::str("Timeout awaiting signal server connection")
+                        .into(),
+                    Sig::new(_sig_drop, ep, sig_uniq, sig_url),
+                ),
+            )
+        })
+        .clone();
+
+    match fut.await {
+        Err(err) => {
+            // if a new sig got added in the mean time, return that instead
+            let r = sig_map.lock().unwrap().get(sig_url).cloned();
+
+            if let Some((new_sig_uniq, new_sig_fut)) = r {
+                if new_sig_uniq != sig_uniq {
+                    return new_sig_fut.await;
+                }
+            }
+
+            Err(err)
+        }
+        Ok(r) => Ok(r),
+    }
+}
+
+fn close_sig(
+    weak_sig_map: &Weak<Mutex<SigMap>>,
+    sig_url: &SigUrl,
+    close_sig_uniq: u64,
+) {
```

Comment:
This is the entrypoint for cleaning up the shared futures in the hash map referenced in my previous comment.